python/cuda_parallel/tests/test_reduce.py

```diff
@@ -550,3 +550,88 @@ def binary_op(x, y):
     d_in = cp.zeros(size)[::2]
     with pytest.raises(ValueError, match="Non-contiguous arrays are not supported."):
         _ = algorithms.reduce_into(d_in, d_out, binary_op, h_init)
+
+
+def test_reduce_with_stream():
+    # Simple cupy stream wrapper that implements the __cuda_stream__ protocol for the purposes of this test
+    class Stream:
+        def __init__(self, cp_stream):
+            self.cp_stream = cp_stream
+
+        @property
+        def __cuda_stream__(self):
+            return (0, self.cp_stream.ptr)
+
+    def add_op(x, y):
+        return x + y
+
+    h_init = np.asarray([0], dtype=np.int32)
+    h_in = random_int(5, np.int32)
+
+    stream = cp.cuda.Stream()
+    with stream:
+        d_in = cp.asarray(h_in)
+        d_out = cp.empty(1, dtype=np.int32)
+
+    stream_wrapper = Stream(stream)
+    reduce_into = algorithms.reduce_into(
+        d_in=d_in, d_out=d_out, op=add_op, h_init=h_init
+    )
+    temp_storage_size = reduce_into(
+        None,
+        d_in=d_in,
+        d_out=d_out,
+        num_items=d_in.size,
+        h_init=h_init,
+        stream=stream_wrapper,
+    )
+    d_temp_storage = cp.empty(temp_storage_size, dtype=np.uint8)
+
+    reduce_into(d_temp_storage, d_in, d_out, d_in.size, h_init, stream=stream_wrapper)
+    np.testing.assert_allclose(d_in.sum().get(), d_out.get())
```

Comment:
We should call `stream.synchronize()` after the call to `reduce_into`. 

Perhaps our wrapper `Stream` type should have a `.synchronize()` method that calls `self.cupy_stream.synchronize()`.

Reply:
Even better is call `cp.testing.assert_allclose(...)` then no need to copy, just need to stream-order it:
```python
with stream:
   cp.testing.assert_allclose(...)
```

Reply:
> it does the copy internally for us, just need to stream-order it:

In general, is it recommended that we rely on this?

Reply:
Yes it's public API.